File: daily/Binary Search/911-Online-election.go
package daily

type TopVotedCandidate struct {
	persons []int
	times   []int
	leaders []int
}

func ConstructorTopVotedCandidate(persons []int, times []int) TopVotedCandidate {
	count := make(map[int]int)
	maxVotes := 0
	leader := -1
	leaders := make([]int, len(persons))
	for i, p := range persons {
		count[p]++
		if count[p] >= maxVotes {
			maxVotes = count[p]
			leader = p
		}
		leaders[i] = leader
	}
	return TopVotedCandidate{
		persons,
		times,
		leaders,
	}
}

func (this *TopVotedCandidate) Q(t int) int {
	lo, hi := 0, len(this.times)-1
	ans := 0
	for lo <= hi {
		mid := (lo + hi) / 2
		if this.times[mid] <= t {
			ans = mid
			lo = mid + 1
		} else {
			hi = mid - 1
		}
	}
	return this.leaders[ans]
}

/**
 * Your TopVotedCandidate object will be instantiated and called as such:
 * obj := Constructor(persons, times);
 * param_1 := obj.Q(t);
 */

File: daily/Binary Search/911-Online-election.py
from bisect import bisect_right
from typing import List


class TopVotedCandidate:
    def __init__(self, persons: List[int], times: List[int]):
        self.times = times
        self.leaders = []

        counts = {}
        leader = -1
        max_votes = 0

        for person in persons:
            counts[person] = counts.get(person, 0) + 1
            if counts[person] >= max_votes:
                max_votes = counts[person]
                leader = person
            self.leaders.append(leader)

    def q(self, t: int) -> int:
        idx = bisect_right(self.times, t) - 1
        if idx < 0:
            return -1
        return self.leaders[idx]
